### ISSUE TITLE
HTTP to HTTPS redirect doesn't work for /login landing page

### ISSUE THREAD

**Issue by Shabirmean** [GoogleCloudPlatform/bank-of-anthos#478]
The `frontend` currently serves requests via a [`redirect`](https://github.com/GoogleCloudPlatform/bank-of-anthos/blob/master/src/frontend/frontend.py#L353) _(I am not completely sure why that is; maybe a Flask thing 🤔 )_, thus regardless of the request URL (`http` or `https`), the response is a `redirect` to a URL constructed with `Flask.url_for()`. 

The url scheme used in `Flask.url_for()` is based on the environment var `SCHEME` as [seen here](https://github.com/GoogleCloudPlatform/bank-of-anthos/blob/master/src/frontend/frontend.py#L517)
```python
app.config['SCHEME'] = os.environ.get('SCHEME', 'http')
```

With this, the way to enable `HTTPS` is to set the environment variable `SCHEME=https` on the `frontend` pod. However still, the initial landing page `GET /login` does not use this redirect format and just renders the `index.html` as [seen here](https://github.com/GoogleCloudPlatform/bank-of-anthos/blob/master/src/frontend/frontend.py#L357). Thus, even with `SCHEME=https` the landing page is served over `http://` if the request is just `http://`.

Whilst this is something that could be handled at the ingress-LB level, this is not ideal to leave it as is in the app. I.E. if the scheme is explicitly set to `https` then the app should always redirect to `https`. Also this can be frictious for cases like Multi Cluster Ingress (MCI) being used whilst MCI still does not supported `http->https` redirects.

**Potential solution fetched from [here](https://stackoverflow.com/questions/45742972/forcing-use-of-https-with-flasks-before-request):**
```python
@app.before_request
def https_redirect():
    if (request.endpoint in app.view_functions and 
         os.environ.get('SCHEME') is 'https' and 
         not request.is_secure):
        return redirect(request.url.replace('http://', 'https://'))
```

**Comment by Shabirmean** [comment]
### TLDR; 
- HTTPS handling should be independent of the applications
- Source code changes would break internal apps that don't use HTTPS to connect to the frontend
- Thus, we will close this issue but have added the following:
     - [PR#482](https://github.com/GoogleCloudPlatform/bank-of-anthos/pull/482) for Single Cluster SSL
     - [PR#483](https://github.com/GoogleCloudPlatform/bank-of-anthos/pull/483) for MultiCluster SSL 
     - We will monitor [b/185580816](b/185580816) for support for HTTP->HTTPS redirect in MultiClusterIngress
---
### Detailed summary

This issue is related to how the BoA frontend application handles requests. Currently the response to the requests to the frontend are redirects. This is probably done to minimize the javascript on the client side and enable page reloads via HTTP redirects.

**So current frontend behaviour:**
- Client _(ex: browser)_ sends an HTTP request
- The application always responds via a HTTP redirect to the next location
- The client sends request to the new HTTP url on the redirect response
- The next page is served to the client _(i.e: browser)_

This is not a problem until the frontend is exposed via an Ingress to enable only HTTPS traffic. When this happens _(if the expectation is to only allow HTTPS)_, then we have to always ensure the redirect urls are “HTTPS” if the request is also “HTTPS”. 

**If this is not done then the behavior becomes:**
- Client _(ex: browser)_ sends an HTTPS request
- The application returns a redirect URL with HTTP as response
- Then the client reaches out to that redirect URL which is HTTP _(not secure anymore)_
- The new page is served over HTTP whereas the initial request was HTTPS

**There are two ways to handle this:**
1. Setup the ingress such that all HTTP requests are redirected to HTTPS at the Ingress itself even before reaching the frontend service
2. Ensure that any response from the frontend _(that is a redirect URL)_ is prefixed with HTTPS

**_Method 1_ (Setup HTTP-HTTPS redirection on Ingress)**
- Ideal solution. 
- Requires no changes in the frontend source code. 
- This should be the way it must be handled where SSL is configured independent of the internal app state
> **Only issues is that this is not supported with MultiClusterIngress (MCI) as of now**

**_Method 2_ (Ensure responses from the frontend are prefixed with HTTPS)** 
- Setting this attribute will ensure that the responses to request have the correct “HTTPS” prefix. 

> **However this will break any internal services (loadgenerator) that invokes the frontend. Because if the redirect URLs are HTTPS, it hasn’t got TLS setup to communicate to the frontend like the browser. Similarly any other applications we add to the system should have TLS setup if the scheme is HTTPS**

_So **Method 2**  cannot be used basically under any circumstance. We don’t necessarily need the SCHEME environment variable. If internal apps need to use TLS then we can use something like istio to enable mTLS via sidecar proxies in all the pods._

Also, since the landing page is using direct Flask `render_template()`, the request to `/login` is going to be HTTP if the requests is also HTTP, even though the scheme is set to HTTPS

---

### Conclusion

The suggested workaround in the issue will not work and will not scale _(i.e: as we add more services we have to also ensure it has an HTTPS mode that is enabled whenever the frontend is HTTPS)_. So the best approach is always to set up HTTPS at a layer above _(independent of)_ the application frontend as with method 1.

However, one unsolved issue with the frontend is that it responds to requests with redirects. This is not nice since, there will be unnecessarily too many requests back and forth. 

**Example scenario of where an Ingress with HTTPS is setup:**
- Client _(ex: browser)_ sends an HTTPS request
- The application returns a redirect URL with HTTP as response
- Then the client reaches out to that redirect URL which is HTTP _(not secure anymore)_
- This HTTP request reaches the Ingress and it redirect HTTP to HTTPS _(Method 1 above)_
- The client receives a HTTP-HTTPS redirect from Ingress
- Now the client reaches out to the frontend on the HTTPS url

![image](https://user-images.githubusercontent.com/7249208/115268090-e09a8c80-a107-11eb-984f-6b7f630fa1d7.png)

_First redirect from frontend and second redirect from Ingress_

We can handle this in a separate improvement independent of this issue. This doesn’t affect in any case what the team wants to achieve with MCI.

**Comment by Shabirmean** [comment]
PR #483 is closed until an update is received for b/185580816
